### PR TITLE
perf(muse): score the packed LM head on the int8 tensor cores (1.11x concurrent decode @c32)

### PR DESCRIPTION
**Pull request by FranDev132**
## Summary

`lm_head` is 202048 x 6656 Q4_K — 756 MB, the single largest matrix in the model, and the last one a
continuous-batch step still scored on the CUDA cores. The dp4a multi-row kernel is chunked by eight
rows, so a 32-row step walked the whole head four times and did its 43 G MAC of dot product on the
wrong units.

The head is wide enough to fill the device on its own: `N/BN` is 6314 blocks against 170 SMs. So
this arm needs neither a K split nor an fp32 scratch, and writes the caller's logits directly — it
is the single-split instantiation of the mma kernel that already scores the attention projections,
with `float` as the output type instead of bf16.

Packed only, which is the restriction the dp4a multi-row arm beside it already carries: its rounding
differs from the rows kernel AR scores with, and a continuous-batch step is a plain argmax emit that
absorbs that while a speculative verify cannot. `decode_packed` sets `packed_rows` and DSpark's
verify never does, so every `dspark-*` path keeps the arm it had.

## Proof of speedup

- [x] Tested on **RTX 5090** (`sm_120`)

**Target model(s)**

- [x] **Muse Glimmer**
- [ ] **Qwen3.8-27B** (ModelOpt NVFP4 / DSpark)
- [ ] **Shared / both** — the change is in code both models use and should help either

Reached only through the packed continuous-batch head at eight rows and up, and only where the head
is Q4_K (`lm_head_type == 12`, which is the branch the call site sits in). Narrower widths and every
non-packed caller keep the dp4a loop, which is also the fallback whenever this arm declines.

**Decode tok/s** (aggregate over 32 concurrent requests, `qwen3_gguf_cb_bench <gguf> 32 256 256 512`):

| | decode tok/s |
|---|--:|
| before | 681.6 |
| after (this PR) | **756.7** |

**Prefill pp tok/s** (not targeted by this PR):

| | prefill pp tok/s |
|---|--:|
| before prefill | — |
| after prefill (this PR) | — |

### Concurrent decode — one binary, `SPARKINFER_HEAD_MMA=0/1`, interleaved

| c | before | after | |
|---|--:|--:|--:|
| 32 | 681.6 | **756.7** | **+11.0%** |
| 16 | 523.1 | **558.7** | **+6.8%** |
| 8 | 321.4 | **328.9** | **+2.3%** |
| 4 | 190.9 | 190.7 | — |
| 2 | 136.9 | 137.2 | — |

Median of three interleaved rounds. Per round:

```text
c=32: before [679.6, 681.6, 719.6]  after [756.1, 756.7, 758.7]
c=16: before [522.9, 523.1, 523.3]  after [557.8, 558.7, 559.1]
c= 8: before [320.9, 321.8]  after [327.2, 328.9, 329.4]
c= 4: before [190.9]  after [190.7]   (below the floor; arm cannot engage)
c= 2: before [136.9]  after [137.2]   (below the floor; arm cannot engage)
```

### Correctness

Not bit-identical, and not claimed to be: the mma path reduces over K in a different order. What a
continuous-batch step actually depends on is the **argmax**, so that is what
`kernels/tests/head_mma_accuracy_gpu_test.cpp` asserts — against `launch_mmvq_rows_f32` at the real
202048 x 6656 shape, at every width this arm serves:

```text
M=32  argmax agree 32/32   relRMS 1.6353e-05   worst/rms 6.9974e-05
M=16  argmax agree 16/16   relRMS 1.3127e-05   worst/rms 4.5772e-05
M=8   argmax agree  8/8    relRMS 1.3677e-05   worst/rms 4.6900e-05
```

The test fails closed if the reference logits are not finite: Q8_1 is `{ __half2 ds; signed char
qs[32] }`, and filling all 36 bytes randomly makes the scale halves inf or NaN, which would have
every logit NaN and both arms' argmax trivially "agree" on index 0.

### CODE REVIEW

**Comment by skyrocket2026** [comment]
## sparkinfer museglimmer auto-eval — `eval-museglimmer:L`

| metric | value |
|---|---|
| **label** | `eval-museglimmer:L` |
| scored at | decode + prefill @ 128/512/4k/16k/32k/64k · concurrent decode @ c2/c4/c8/c16/c32 — 17 axes, each also a regression floor; the label is the best |
| tier from | `muse-cb-decode@c32` (+12.6%) |
| accuracy gate | ✅ top1=0.970 (bar >=0.9) · KL=0.0906 (bar <=0.1) |
| qwen3.6 guard | ✅ no regression (decode+prefill @ 32k) |
| modelopt guard | ✅ no regression (decode+prefill @ 32k, Qwen3.8-27B NVFP4) |
| PPL sparkinfer / llama.cpp | 2.6881 / 3.0876 |
| Polaris receipt | collected, not signed (no key configured) |
| commit | `5b8e54106` |

| ctx | phase | main | PR | delta |
|---|---|---|---|---|
| 128 | decode | 106.60 | 106.66 | +0.1% |
| 128 | prefill | 9852.88 | 9848.68 | -0.0% |
| 512 | decode | 105.66 | 105.79 | +0.1% |
| 512 | prefill | 15919.84 | 15986.46 | +0.4% |
| 4k | decode | 102.75 | 102.80 | +0.1% |
| 4k | prefill | 14937.36 | 14944.73 | +0.0% |
| 16k | decode | 101.80 | 101.82 | +0.0% |
| 16k | prefill | 14367.27 | 14395.49 | +0.2% |
| 32k | decode | 100.41 | 100.49 | +0.1% |
| 32k | prefill | 13084.31 | 13032.66 | -0.4% |
| 64k | decode | 98.28 | 98.25 | -0.0% |
| 64k | prefill | 11210.74 | 11205.28 | -0.0% |

**Concurrent decode** — aggregate tok/s with N requests in flight

| concurrency | main | PR | delta |
|---|---|---|---|
| c2 | 135.90 | 137.00 | +0.8% |
| c4 | 190.50 | 190.50 | +0.0% |
| c8 | 321.30 | 329.00 | +2.4% |
| c16 | 522.60 | 559.60 | +7.1% |
| c32 | 681.50 | 767.50 | +12.6% |

ok

<sub>Scored on the pinned eval box vs same-box `origin/main` — AR decode AND prefill at ctx 128/512/4k/16k/32k/64k, plus concurrent decode (aggregate tok/s) at c2/c4/c8/c16/c32; ANY axis regressing is a hard REJECT, but otherwise the reported label is the **best** measured delta across all 17 — a PR that improves just one, with the rest flat, still earns credit for that. Cross-model no-regression guards run at 32k on Qwen3.6 and the ModelOpt Qwen3.8-27B NVFP4 checkpoint. This is informational, not a judgment on your PR: a `none` label just means no measurable Muse Glimmer speedup was verified on either metric, which is expected and fine if that isn't what your change is about. Correctness gated against a live llama.cpp reference on the same GGUF. Also gated on two cross-model no-regression guards at 32k (decode+prefill, same box vs main): Qwen3.6-35B-A3B and the ModelOpt Qwen3.8-27B NVFP4 checkpoint — Muse Glimmer PRs can touch code shared with other models. Automated. The round's best-scoring PR may be auto-merged as `merge-first` once every gate above passes; a separate comment says so explicitly when that happens.</sub>

**Comment by skyrocket2026** [comment]
Auto-merged as the round's `museglimmer-merge-first` winner — verified same-box 128-token decode speedup over `main`, accuracy-gated vs llama.cpp.
